src/notebooks/controllers/liveKernelSwitcher.ts

```diff
@@ -23,56 +22,54 @@ const MEMENTO_BASE_KEY = 'jupyter-notebook-remote-session-';
     }
 
     private onDidOpenNotebook(n: NotebookDocument) {
-        const key = this.getKey(n.uri);
-        const session = this.memento.get(key);
-        if (session) {
-            // When all controllers are loaded, see if one matches
-            this.controllerManager.kernelConnections
-                .then(async (list) => {
-                    const active = this.controllerManager.getSelectedNotebookController(n);
-                    const matching = list.find((l) => l.id === session);
-                    if (matching && active?.id !== matching.id) {
-                        // This controller is the one we want, but it's not currently set.
-                        await switchKernel(n.uri, this.vscNotebook, undefined, this.commandManager, matching);
-                    } else {
-                        // There's no match, so live connection must be gone
-                        await this.memento.update(key, undefined);
-                    }
-                })
-                .catch((e) => traceError(e));
-        }
+        // When all controllers are loaded, see if one matches
+        this.controllerManager.kernelConnections
+            .then(async (list) => {
+                const active = this.controllerManager.getSelectedNotebookController(n);
+                const preferredRemote = this.preferredRemoteKernelIdProvider.getPreferredRemoteKernelId(n.uri);
+                const matching = preferredRemote && list.find((l) => l.id === preferredRemote);
+                if (matching && active?.id !== matching.id) {
+                    traceInfo(`Switching remote kernel to ${preferredRemote} for ${n.uri}`);
+                    // This controller is the one we want, but it's not currently set.
+                    await this.switchKernel(n, matching);
+                }
+            })
+            .catch((e) => traceError(e));
     }
 
-    private onLiveRefresh(liveConnections: LiveRemoteKernelConnectionMetadata[]) {
-        // When a refresh happens, remember the live connection id for all notebooks
-        this.vscNotebook.notebookDocuments.forEach(async (n) => {
-            const kernel = this.kernelProvider.get(n.uri);
-            if (kernel) {
-                const match = liveConnections.find((c) => c.kernelModel.id === kernel.session?.kernelId);
-                if (match) {
-                    const key = this.getKey(n.uri);
-                    await this.memento.update(key, match.id);
+    private async switchKernel(n: NotebookDocument, kernel: Readonly<KernelConnectionMetadata>) {
+        traceInfo(`Using notebook.selectKernel to force remote kernel for ${n.uri} to ${kernel.id}`);
+        // Do this in a loop as it may fail
+        const success = await waitForCondition(
+            async () => {
+                if (this.vscNotebook.activeNotebookEditor?.document === n) {
+                    await this.commandManager.executeCommand('notebook.selectKernel', {
```

Comment:
first time using this command in production.

Reply:
Not true. KernelSelector.ts has it already. I moved this here to be more explicit about the loop.

Reply:
I thought that was specific to Interactive Window with different arguments.
Doesn't matter anyways.

Reply:
It was merely an observation.